### PR TITLE
Slip stream in composer 2.0

### LINKED ISSUE #13: Adds support for Composer v2
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| BC Break    | maybe ?

#### Summary

Composer v2 is behind the corner and we should be ready on the day 0, so we need updte this library ASAP.

This PR might be helpful:
https://github.com/Ocramius/PackageVersions/pull/133

### PR DESCRIPTION
**Pull request by boesing**
|    Q          |   A
|-------------- | ------
| New Feature   | yes

### Description

While trying to migrate to composer 2.0, I've tried many ways on how to achieve the functionality of the current plugin.

The current plugin "slip streams" into the composer process, which is probably the best solution to solve the following scenarios:

- If 3rd-party library has a zend-dependency and updates to laminas, there wont be any changes to the project as the laminas package is already required.

- If a 3rd-party library has a zend-dependency and requires a newer zend-dependency, the project will just upgrade to the laminas equivalent. No further action needed.

- If a 3rd-party library totally drops a zend-dependency and does not need the laminas equivalent, the slip-streamed laminas dependency will automatically being removed from the project if its not needed anymore.

I came up with a way in #15 which should have introduced some more user interactions. The main idea was:

If a 3rd-party library has a zend-dependency, ask the user if he wants to replace that zend-dependency the laminas pendant, write the dependency to `composer.json` of the project and remember that we introduced that dependency due to a non-migrated library.

### TL;DR
This is finally ready for review.

### Story on why I decided to go this way

After many sleepless nights and a total rewrite of the dependency plugin, I realized that this approach would probably be a future proof solution but its way too complex.
The complexity exceeded my brain capacity too often.
Its unmaintainable due to all those possible combinations:
- 3rd-party library no.1 uses zend-dependency and is required via root `composer.json` as a `require` package
- 3rd party library no.2 uses same zend-dependency and is required via root `composer.json` as a `require-dev` package
- 3rd party library no.1 is being uninstalled
- root `composer.json` should change the `require` of the laminas equivalent to `require-dev` as there is only a dev-requirement left which requires the zend-package

And thats just one of those complex scenarios... 

However, after I realized that #15 wont be a solution I am happy with, I started to work on this approach today. I don't know why I tried this earlier (as I was debugging what exactly changed in composer 2.0 that the slip-stream solution we are actually use does not work anymore)... That would probably saved my a ton of work.

### CODE REVIEW

File: src/AbstractDependencyRewriter.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    abstract public function onPrePackageInstallOrUpdate(PackageEvent $event);
+
+    /**
+     * @param string $message
+     * @param int $verbosity
+     */
+    protected function output($message, $verbosity = IOInterface::NORMAL)
+    {
+        $this->io->write($message, $newline = true, $verbosity);
```
Comment (samsonasik) [inline src/AbstractDependencyRewriter.php]: `$newline` var is no need to be defined ? should can be: 

```suggestion
        $this->io->write($message, true, $verbosity);
```

File: src/AbstractDependencyRewriter.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            sprintf(
+                'Changing package in current command from %s to %s',
+                $name,
+                $replacementName
+            ),
+            true,
+            IOInterface::DEBUG
+        );
+
+        $version = count($result) ? array_shift($result) : null;
```
Comment (samsonasik) [inline src/AbstractDependencyRewriter.php]: no need count, can use truthy check:

```suggestion
        $version = $result ? array_shift($result) : null;
```

File: src/DependencyRewriterV1.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    /**
+     * @param object $object
+     * @param string $property
+     * @param mixed $value
+     */
+    private function updateProperty($object, $property, $value)
+    {
+        $r = new ReflectionProperty($object, $property);
+        $r->setAccessible(true);
+        $r->setValue($object, $value);
```
Comment (samsonasik) [inline src/DependencyRewriterV1.php]: can use closure bind, something like:

```php
    (function ($object, $property, $value) {
        $object->$property = $value;
    })->bindTo($object, $object)($object, $property, $value);
```

ref https://ocramius.github.io/blog/accessing-private-php-class-members-without-reflection/

File: test/DependencyRewriterPluginDelegatorTest.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        $rewriter = $this->createMock($rewriter);
+        $rewriter
+            ->expects($this->once())
+            ->method($eventMethod)
+            ->with($event);
+
+        self::assertInstanceOf(RewriterInterface::class, $rewriter);
+        $delegator = new DependencyRewriterPluginDelegator($rewriter);
+
+        call_user_func([$delegator, $eventMethod], $event);
```
Comment (samsonasik) [inline test/DependencyRewriterPluginDelegatorTest.php]: should can use : 

```suggestion
        $delegator->$eventMethod($event);
```

Comment (boesing) [reply test/DependencyRewriterPluginDelegatorTest.php]: Just curious, but where is the benefit with this?
Is it less opcodes?

File: src/DependencyRewriterV1.php (hunk 3 above)
Comment (boesing) [reply src/DependencyRewriterV1.php]: This is all copy & pasted from current version.
I know the closure thing but didnt want to change current behavior.

File: src/AbstractDependencyRewriter.php (hunk 2 above)
Comment (boesing) [reply src/AbstractDependencyRewriter.php]: Same here, this is from current behavior. Just copy & pasted to new class

File: src/AbstractDependencyRewriter.php (hunk 1 above)
Comment (boesing) [reply src/AbstractDependencyRewriter.php]: Copy & paste from current code

File: test/DependencyRewriterPluginDelegatorTest.php (hunk 4 above)
Comment (samsonasik) [reply test/DependencyRewriterPluginDelegatorTest.php]: mostly less code, no need `use function call_user_func`  in the header.

Comment (boesing) [reply test/DependencyRewriterPluginDelegatorTest.php]: I'll change that, even if its non-productive code and thus there should be no need for that kind of micro optimization imho.

**Comment by boesing** [comment]
Well, now travis fails because codestandard says "closure can be static" but it can't as it wont set the property anymore... Thats why I hate changing code I didn't wrote myself... 👎

**Comment by samsonasik** [comment]
I didn't write "static" in my example https://github.com/laminas/laminas-dependency-plugin/pull/18#discussion_r429628900 , you added static yourself, that's why it got error.

**Comment by boesing** [comment]
> I didn't write "static" in my example [#18 (comment)](https://github.com/laminas/laminas-dependency-plugin/pull/18#discussion_r429628900) , you added static yourself, that's why it got error.

I've applied phpcbf which changed it to `static` without me noticing.
After tests failed, I dropped static and now phpcs is arguing about missing `static`.

Welcome to QA hell.

**Comment by boesing** [comment]
@weierophinney Any ideas how to raise code coverage? it needs to get merged from both results of `COMPOSER_VERSION` v1 and v2...

**Comment by weierophinney** [comment]
> After tests failed, I dropped static and now phpcs is arguing about missing `static`.
> 
> Welcome to QA hell.

Wrap the closure (or at least its opening line) with `// phpcs:disable` and `// phpcs:enable`, if tests fail when it is marked as static.

**Comment by weierophinney** [comment]
> Any ideas how to raise code coverage? it needs to get merged from both results of `COMPOSER_VERSION` v1 and v2...

No clues whatsoever... maybe @michalbundyra might have ideas?

File: composer.json [hunk 5]
```diff
@@ -8,15 +8,14 @@
...
     "require": {
         "php": "^5.6 || ^7.0",
-        "composer-plugin-api": "^1.1"
+        "composer-plugin-api": "^1.1 || ^2.0"
     },
     "require-dev": {
-        "composer/composer": "^1.9",
-        "dealerdirect/phpcodesniffer-composer-installer": "^0.5.0",
+        "composer/composer": "^1.9 || ^2.0@dev",
+        "composer/semver": "^1.5 || ^2.1@dev",
```
Comment (michalbundyra) [inline composer.json]: We should be able to change it - now version 3 is released, so I believe we can remove this requirement completely.

Comment (boesing) [reply composer.json]: Feel free to do that in another PR. I will do nothing until this is merged. There are hundrets of dependency changes every day. Next week might be v4 and we are already merged.
This PR frustrated me at a level I cannot even describe.

Comment (boesing) [reply composer.json]: Done... I hope composer wont use another `@dev` in one of the next commits...

**Comment by boesing** [comment]
Well, I finally got proper codecoverage.
Its around 92% now.

Hopefully, thats enough. I also removed dev-dependency of `composer/semver` and thus, I think we're ready for a release candidate.

**Review by samsonasik — APPROVED** [review]

File: phpcs.xml [hunk 6]
```diff
@@ -1,14 +1,18 @@
 <?xml version="1.0"?>
 <ruleset name="Webimpress Coding Standard">
-    <rule ref="./vendor/webimpress/coding-standard/ruleset.xml"/>
+    <rule ref="./vendor/webimpress/coding-standard/ruleset.xml">
```
Comment (weierophinney) [inline phpcs.xml:3]: We should probably move to Laminas Coding Standard v2, but that can wait for another PR. :smile:

**Review by weierophinney — APPROVED** [review]
This looks fantastic - surprised to see you found a way to keep the current functionality with the v2 API - I'm assuming the "POOL" events must be the secret to making that work?

I have no feedback at a technical level. I haven't had time to dive into where and when the POOL events are triggered, but I trust _you_ did the research!

Let's do an alpha or beta release from this, and have people start testing!

**Comment by boesing** [comment]
> I'm assuming the "POOL" events must be the secret to making that work?

Yah, more or less. Jordi pointed me to this event, but it is being triggered before any package is being downloaded. As I've started to implement the initial version, I realized, that the dependency-plugin was not able to interact due to the fact, that it was not downloaded at the point when the events are triggered (in a project which is being installed for the first time - so without the vendor/ directory available).

That was what made me thinking about the way we discussed a few months ago. That other way drove me crazy after a few weeks of developing and thus I tried this approach again.
However, I am happy that this works, even if I dont like the fact that we have to re-run composer again to ensure we have our packages slip-streamed.

Flow is as follows (for a freshly installed package):
- Pool event is being triggered
- Packages are being downloaded to vendor/
- Packages are being triggered (install or update event)
- Plugin detects zend packages
- Composer executes dump-autoloader
- Plugin hooks into that and executes `uninstall` for all zend packages
- Plugin executes `composer update --lock`
- Pool event is being triggered
- Plugin changes zend packages to laminas packages in the pool
- composer.lock is being written with the laminas replacements
- Composer installs laminas package in favor of zend package
- dump-autoload is being triggered again (this time we dont care as we did our job)

So hopefully, we have the same features as in the first version of this plugin.

@weierophinney I would love to see this as a first preview release so contributors can start testing with the new version.

**Comment by weierophinney** [comment]
@boesing Go ahead and merge to develop, and issue a 2.0.0beta1 release. I'm suggesting beta, in case we determine there might be additional features before we do a stable release; that way we can do another beta. Otherwise, we can jump directly from beta to stable if user tests show it working as expected.

**Comment by roelvanduijnhoven** [comment]
Just want to give feedback: works great for me with the composer 2.0 plugin :). Just pushed our app to production, using this beta.

File: composer.json [hunk 7]
```diff
@@ -8,15 +8,13 @@
     },
     "require": {
         "php": "^5.6 || ^7.0",
-        "composer-plugin-api": "^1.1"
+        "composer-plugin-api": "^1.1 || ^2.0"
     },
     "require-dev": {
-        "composer/composer": "^1.9",
-        "dealerdirect/phpcodesniffer-composer-installer": "^0.5.0",
+        "composer/composer": "^1.9 || ~2.0.0@dev || ^2.0",
```
Comment (snapshotpl) [inline composer.json:14]: `~2.0.0@dev` can be removed after stable 2.0

File: src/AbstractDependencyRewriter.php [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+use function array_shift;
+use function get_class;
+use function in_array;
+use function preg_split;
+use function reset;
+use function sprintf;
+
+abstract class AbstractDependencyRewriter implements RewriterInterface
+{
+    /** @var Composer */
```
Comment (snapshotpl) [inline src/AbstractDependencyRewriter.php:26]: Can be `null`

File: src/AbstractDependencyRewriter.php [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    protected $composer;
+
+    /** @var IOInterface */
```
Comment (snapshotpl) [inline src/AbstractDependencyRewriter.php:29]: Can be `null`
